src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,531 @@
+    assume t ht,
+    simp only [exists_prop, mem_Union, mem_singleton_iff] at ht,
+    rcases ht with ⟨e', e'_atlas, s, s_open, ts⟩,
+    rw ts,
+    let f := e.symm.trans e',
+    have : is_open (f.to_fun ⁻¹' s ∩ f.source),
+      by simpa [inter_comm] using (continuous_on_open_iff (c.open_source e e' he e'_atlas)).1
+        (c.continuous_to_fun e e' he e'_atlas) s s_open,
+    have A : e'.to_fun ∘ e.inv_fun ⁻¹' s ∩ (e.target ∩ e.inv_fun ⁻¹' e'.source) =
+             e.target ∩ (e'.to_fun ∘ e.inv_fun ⁻¹' s ∩ e.inv_fun ⁻¹' e'.source),
+      by { rw [← inter_assoc, ← inter_assoc], congr' 1, exact inter_comm _ _ },
+    simpa [local_equiv.trans_source, preimage_inter, preimage_comp.symm, A] using this
+  end,
+  ..e }
+
+def to_manifold : @manifold H _ M c.to_topological_space :=
+{ atlas := ⋃ (e : local_equiv M H) (he : e ∈ c.atlas), {c.local_homeomorph e he},
+  chart_at := λx, c.local_homeomorph (c.chart_at x) (c.chart_mem_atlas x),
+  mem_chart_source := λx, c.mem_chart_source x,
+  chart_mem_atlas := λx, begin
+    simp only [mem_Union, mem_singleton_iff],
+    exact ⟨c.chart_at x, c.chart_mem_atlas x, rfl⟩,
+  end }
+
+end manifold_core
+
+section has_groupoid
+
+/-- A manifold has an atlas in a groupoid G if the change of coordinates belong to the groupoid -/
+class has_groupoid (H : Type*) [topological_space H] (M : Type*) [topological_space M]
+  [manifold H M] (G : structure_groupoid H) : Prop :=
+(compatible : ∀e e' : local_homeomorph M H, e ∈ atlas M H → e' ∈ atlas M H → (e.symm →ₕ e') ∈ G)
+
+lemma has_groupoid_of_le {H : Type*} [topological_space H] {M : Type*} [topological_space M]
+  [manifold H M] {G₁ G₂ : structure_groupoid H} (h : has_groupoid H M G₁) (hle : G₁ ≤ G₂) :
+  has_groupoid H M G₂ :=
+⟨ λ e e' he he', hle ((h.compatible : _) e e' he he') ⟩
+
+/-- The trivial manifold structure on the model space is compatible with any groupoid -/
+instance has_groupoid_model_space (H : Type*) [topological_space H] (G : structure_groupoid H) :
+  has_groupoid H H G :=
+{ compatible := λe e' he he', begin
+    replace he : e ∈ atlas H H := he,
+    replace he' : e' ∈ atlas H H := he',
+    rw model_space_atlas at he he',
+    simp [he, he', structure_groupoid.id_mem]
+  end }
+
+variable [topological_space H]
+
+/-- In a manifold having some structure groupoid, the changes of coordinates belong to this groupoid -/
+lemma compatible (G : structure_groupoid H) {M : Type*} [topological_space M] [manifold H M]
+  [h : has_groupoid H M G]
+  {e e' : local_homeomorph M H} (he : e ∈ atlas M H) (he' : e' ∈ atlas M H) :
+  e.symm →ₕ e' ∈ G :=
+(h.compatible : _) e e' he he'
+
+/-- A G-diffeomorphism between two manifolds is a homeomorphism which, when read in the charts,
+belongs to G. -/
+structure diffeomorph (G : structure_groupoid H) (M : Type*) (M' : Type*)
+  [topological_space M] [topological_space M'] [manifold H M] [manifold H M']
+  extends homeomorph M M' :=
+(smooth_to_fun : ∀c : local_homeomorph M H, ∀c' : local_homeomorph M' H,
```

Comment:
I don't know if it works, but it would be cool if instead of `∀c : local_homeomorph M H,` you could just immediately write `∀c ∈ atlas M H`. I'm not sure what the exact conditions are for this syntactic sugar to work.

Reply:
Alas, it doesn't work. It does not even work to omit the type of `c`, which is strange since Lean should be able to deduce it from the line `c ∈ atlas M H`. Well, unification is complicated, as we know.

Reply:
Too bad. Then we'll have to deal with an extra line of code (-;